web/src/settings_users.js

```diff
@@ -94,6 +113,52 @@ function failed_listing_users() {
     blueslip.error("Error while listing users for user_id", {user_id});
 }
 
+function add_value_to_filters(section, key, value) {
+    section.filters[key] = value;
+    // This hard_redraw will rerun the relevant predicate function
+    // and in turn apply the new filters.
+    section.list_widget.hard_redraw();
+}
+
+function role_selected_handler(event, dropdown, widget) {
+    event.preventDefault();
+    event.stopPropagation();
+
+    const role_code = Number($(event.currentTarget).attr("data-unique-id"));
+    if (widget.widget_name === section.active.dropdown_widget_name) {
+        add_value_to_filters(section.active, "role_code", role_code);
+    } else if (widget.widget_name === section.deactivated.dropdown_widget_name) {
+        add_value_to_filters(section.deactivated, "role_code", role_code);
+    }
+
+    dropdown.hide();
+    widget.render();
+}
+
+function get_roles() {
+    return [
+        {unique_id: "0", name: $t({defaultMessage: "All roles"})},
+        ...Object.values(settings_config.user_role_values).map((user_role_value) => ({
+            unique_id: user_role_value.code.toString(),
+            name: user_role_value.description,
+        })),
+    ];
+}
+
+function create_role_filter_dropdown($events_container, widget_name) {
+    new dropdown_widget.DropdownWidget({
+        widget_name,
+        get_options: get_roles,
+        $events_container,
+        item_click_callback: role_selected_handler,
+        default_id: "0",
+        tippy_props: {
+            placement: "bottom-start",
+            offset: [0, 0],
+        },
+    }).setup();
+}
```

Comment:
Having a `section` parameter and defining `role_selected_handler` inside this function might let us clean up `role_selected_handler` to not need the conditionals ... but I don't think it's worth reworking given that we intended to merge the panels anyway.